Created ability to configure allowed URL in config.xml.

millb opened the pull request:
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):
User can add in `<remote_url_allow_hosts>` from `config.xml`  hosts, which other users can go to.
If `<remote_url_allow_hosts>` doesn't exist or it's empty then nothing changes.

millb committed:
Created ability to configure allowed URL in config.xml without tests.

alesapin commented on dbms/src/Storages/StorageURL.cpp:
```diff
@@ -15,13 +15,17 @@
 #include <Poco/Net/HTTPRequest.h>
+#include <Poco/NumberFormatter.h>
+
+#include <../../contrib/re2/re2/re2.h>
```
Very strange import :) https://github.com/clickhouse/ClickHouse/blob/master/dbms/src/Parsers/ASTColumnsMatcher.cpp#L7

alesapin commented on dbms/src/Storages/StorageURL.cpp:
```diff
@@ -34,6 +38,9 @@ IStorageURLBase::IStorageURLBase(
 {
+    if (!checkHostWhitelist(uri.getHost()) &&
+        !checkHostWhitelist(uri.getHost() + ":" + Poco::NumberFormatter::format(uri.getPort())))
+        throw Exception("Unacceptable URL.", ErrorCodes::UNACCEPTABLE_URL);
```
Very unclear message.

alesapin commented on dbms/src/Storages/StorageURL.cpp:
```diff
@@ -221,4 +228,22 @@ void registerStorageURL(StorageFactory & factory)
+bool IStorageURLBase::checkHostWhitelist(const std::string & host) {
+    const std::unordered_set<std::string> primary_hosts = context_global.getAllowedPrimaryUrlHosts();
+    const std::vector<std::string> regexp_hosts = context_global.getAllowedRegexpUrlHosts();
+    if (!(primary_hosts.empty() && regexp_hosts.empty()))
```
Use De Morgan laws https://en.wikipedia.org/wiki/De_Morgan%27s_laws

alesapin commented on dbms/src/Storages/StorageURL.cpp:
```diff
@@ -221,4 +228,22 @@ void registerStorageURL(StorageFactory & factory)
+            bool flag = false;
+            for (size_t i = 0; i < regexp_hosts.size() && !flag; ++i)
+                if (re2::RE2::FullMatch(host, regexp_hosts[i]))
+                    flag = true;
```
Why we don't exit the loop right here?

millb committed:
Fixed StorageUrl.cpp

millb committed:
Fixed StorageURL.cpp

millb committed:
Add integration tests.

millb committed:
Fixed test codestyle

millb committed:
Fixed codestyle and changed Error message.

millb committed:
Fixed tests

alesapin commented on dbms/src/Storages/StorageURL.cpp:
```diff
@@ -34,6 +38,11 @@ IStorageURLBase::IStorageURLBase(
     : uri(uri_), context_global(context_), format_name(format_name_), table_name(table_name_), database_name(database_name_)
 {
+    if (!checkHost(uri.getHost()) &&
+        !checkHost(uri.getHost() + ":" + Poco::NumberFormatter::format(uri.getPort())))
```
`toString` (std::to_string).

alexey-milovidov commented on dbms/programs/server/Server.cpp:
```diff
@@ -507,6 +507,23 @@ int Server::main(const std::vector<std::string> & /*args*/)
+            else if (startsWith(key, "host"))
+                primary_hosts.insert(config().getString("remote_url_allow_hosts." + key));
+        }
+        global_context->setAllowedPrimaryUrlHosts(primary_hosts);
```
We prefer URL, not Url.

alexey-milovidov commented on dbms/programs/server/config.xml:
```diff
@@ -3,6 +3,25 @@
+			Example: "yandex.ru", "yandex.ru." and "www.yandex.ru" are different hosts.
+            If port is explicitly specified in URL, the host:port is checked as a whole.
+            If host specified here without port, any port with this host allowed.
+            "yandex.ru" -> "yandex.ru:443", "yandex.ru:80" ets. is allowed, but "yandex.ru:80" -> only "yandex.ru:80" is allowed. 
```
ets -> etc.

alexey-milovidov commented on dbms/src/Interpreters/Context.cpp:
```diff
@@ -156,7 +156,8 @@ struct ContextShared
 
     std::unique_ptr<TraceCollector> trace_collector;        /// Thread collecting traces from threads executing queries
-
+    std::unordered_set<std::string> allowed_primary_url_hosts;      /// Allowed primary (<host>) URL from config.xml
```
You can encapsulate it into separate class that will be responsible for:
- initializing from configuration;
- storing the set of allowed hosts;
- checking if the host is allowed;
- throwing exceptions.

It's much better than if it's spread across the codebase.

alexey-milovidov commented on dbms/tests/integration/test_allowed_url_from_config/configs/config_with_only_primary_hosts.xml:
```diff
@@ -0,0 +1,10 @@
+        <host>host:123</host>
+        <host>yandex.ru</host>
+
+	</remote_url_allow_hosts>
```
tab

alexey-milovidov commented on dbms/tests/integration/test_allowed_url_from_config/configs/config_with_only_regexp_hosts.xml:
```diff
@@ -0,0 +1,9 @@
+<yandex>
+	<remote_url_allow_hosts>
```
.

alexey-milovidov commented on dbms/tests/integration/test_allowed_url_from_config/test.py:
```diff
@@ -0,0 +1,52 @@
+    assert node5.query("CREATE TABLE table_test_5_1 (word String) Engine=URL('https://host:80', CSV)") == ""
+    assert node5.query("CREATE TABLE table_test_5_2 (word String) Engine=URL('https://host', CSV)") == ""
+    assert node5.query("CREATE TABLE table_test_5_3 (word String) Engine=URL('https://yandex.ru', CSV)") == ""
+    assert node5.query("CREATE TABLE table_test_5_4 (word String) Engine=URL('ftp://something.com', CSV)") == ""
```
It should also be checked for `remote`, `remoteSecure`, `s3` and `hdfs`.

alexey-milovidov commented on dbms/src/Storages/StorageURL.cpp:
```diff
@@ -34,6 +38,11 @@ IStorageURLBase::IStorageURLBase(
     const ConstraintsDescription & constraints_)
     : uri(uri_), context_global(context_), format_name(format_name_), table_name(table_name_), database_name(database_name_)
 {
+    if (!checkHost(uri.getHost()) &&
```
Cannot find the check for redirects.

millb committed:
Fixed Error message and error code

millb committed:
Created class StorageOfAllowedURL
Created test for table function remote and remoteSecure
Created check allowed url in remote and remoteSecure
Fixed tests

millb committed:
Update config.xml

millb committed:
Update StorageOfAllowedURL.cpp

millb committed:
Update StorageOfAllowedURL.h

millb committed:
Update Context.cpp

millb committed:
Update TableFunctionRemote.cpp

millb committed:
Update TableFunctionRemote.cpp

millb committed:
Renamed StorageOfAllowedURL -> RemoteHostFilter

millb committed:
Delete StorageOfAllowedURL.cpp

millb committed:
Delete StorageOfAllowedURL.h

millb committed:
Merge remote-tracking branch 'origin/url_engine_problem' into url_engine_problem

millb committed:
Update RemoteHostFilter.h

millb committed:
Update RemoteHostFilter.cpp

millb committed:
Merge branch 'master' into url_engine_problem

millb committed:
Redirect support added

millb committed:
Support redirect improvement

millb committed:
Support redirect improvement

millb committed:
Merge branch 'url_engine_problem' of github.com:millb/ClickHouse into url_engine_problem

millb committed:
Update ReadWriteBufferFromHTTPWithHostFilter.cpp

millb committed:
Fixed bugs

millb committed:
Delete ReadWriteBufferFromHTTPWithHostFilter.cpp

millb committed:
Delete ReadWriteBufferFromHTTPWithHostFilter.h

millb committed:
Update config_with_only_primary_hosts.xml

millb committed:
Update config_with_only_regexp_hosts.xml

millb committed:
Update config_without_allowed_hosts.xml

millb committed:
Update config_with_hosts.xml

millb committed:
Update config_with_hosts.xml

millb committed:
Update config_with_only_regexp_hosts.xml

millb committed:
Update config_without_allowed_hosts.xml

millb committed:
fixed_merge_conflict:wq

millb committed:
Update ErrorCodes.cpp

millb committed:
Merge branch 'master' into url_engine_problem

millb commented:
Stress test (thread) fail isn't caused by my changes

alesapin commented on dbms/src/Common/RemoteHostFilter.h:
```diff
@@ -0,0 +1,28 @@
+class RemoteHostFilter
+{
+public:
+    void checkURL(const Poco::URI &uri) const; /// If URL not allowed in config.xml throw UNACCEPTABLE_URL Exception
```
Style of `&`.

alesapin reviewed (commented):
Also, we need to add filters for StorageHDFS and StorageS3.

millb commented on dbms/src/Common/RemoteHostFilter.cpp:
```diff
@@ -0,0 +1,62 @@
+
+bool RemoteHostFilter::checkString(const std::string &host) const
+{
+    if (!primary_hosts.empty() || !regexp_hosts.empty())
```
If primary_hosts and regexp_hosts are empty, it will mean that no host and regexp_host tegs in config.xml. In this case all hosts allowed.

millb committed:
Merge branch 'master' into url_engine_problem

millb committed:
Merge remote-tracking branch 'refs/remotes/origin/url_engine_problem' into url_engine_problem

millb committed:
Created check for HDFS and S3 storages
:

Commit:
working commit

Commit:
Merge branch 'url_engine_problem' of https://github.com/millb/ClickHouse into url_engine_problem

Commit:
Created test for HDFS Storage and hdfs table function
Deleted extra tests

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into url_engine_problem

millb committed:
Working commit

Commit:
Merge branch 'url_engine_problem' of https://github.com/millb/ClickHouse into url_engine_problem

Commit:
Merge branch 'master' into url_engine_problem

Commit:
Merge branch 'master' into url_engine_problem

Commit:
Added integration test for storage_s3

millb committed:
Update WriteBufferFromS3.cpp

millb committed:
Update ReadBufferFromS3.cpp

millb committed:
Merge branch 'master' into url_engine_problem

millb committed:
Merge branch 'url_engine_problem' of github.com:millb/ClickHouse into url_engine_problem

Commit:
Merge branch 'master' of https://github.com/ClickHouse/ClickHouse into url_engine_problem

Hipska commented:
The example config does not show how the `remote_url_allow_hosts` items should look like. Also nothing about this can be found in the documentation.
